### PR TITLE
Add Localization for the EnKF

### LINKED ISSUE #269: Implement Localization for EnKF
There are two tricks that are extremely common in EnKF applications: covariance inflation and localization. We currently have the former, but not the latter, which is crucial for application of the EnKF in many high-dimensional problems. 

A nice start would be including the Gaspari-Cohn covariance tapering approach, as discussed in [Katzfuss et al., 2016](https://www.math.umd.edu/~slud/RITF17/enkf-tutorial.pdf).

### PR DESCRIPTION
**Pull request by DanWaxman**
This PR adds localization via covariance tapering, namely through the Gaspari-Cohn and Gaussian correlation functions. 

- The proposed API adds a `GetCovarianceTapers` callback to the filtering interface, which specifies a cross-covariance taper and, optionally, a marginal covariance taper. 
- Custom tapers are possible, but we also provide Gaspari-Cohn and Gaussian in `cuthbertlib/ensemble_kalman/localization.py`.
  - Gaspari-Cohn is more standard, but Gaussian has infinite support, and therefore has better gradient properties if one wants to perform gradient-based optimization of localization hyperparameters.
- We add a tutorial using Lorenz-96.
- We have end-to-end tests to show (i) localization doesn't crash things, basically and (ii) gradients w.r.t. lengthscale of a Gaussian taper are well-behaved, matching closely to a finite difference approximation. 

We purposefully avoid localization for smoothing, as there lacks a straightforward "correct" way to do so, as far as I can tell.

### CODE REVIEW

File: cuthbertlib/ensemble_kalman/filtering.py [hunk 1]
```diff
@@ -95,11 +107,19 @@ def update(
     x_dev = predicted_ensemble - x_mean
     y_dev = y_pred - y_mean
 
-    # Square-root innovation covariance via tria
-    chol_S = tria(jnp.concatenate([y_dev.T / jnp.sqrt(N - 1), chol_R], axis=1))
-
     # Cross-covariance
     C_xy = x_dev.T @ y_dev / (N - 1)
+    if tapers is not None:
+        C_xy = tapers.cross * C_xy
```
Comment (SamDuffield) [inline cuthbertlib/ensemble_kalman/filtering.py]: I'm wondering if the tapering API could be a function that modifies the cross or marginal covariance rather than an array itself (the current `CovarianceTapers` object). This would be more flexible to the user and allow tapering that doesn't require instantiating an extra d x d matrix. The only issue I see is I'm not sure it will play nice with the argsort/nan code above

Comment (SamDuffield) [reply cuthbertlib/ensemble_kalman/filtering.py]: I.e. `modified_cross, modified_marginal = covariance_modify(cross, marginal, model_inputs)`

what do you think?

File: cuthbertlib/ensemble_kalman/filtering.py [hunk 2]
```diff
@@ -95,11 +107,19 @@ def update(
...
+
+    if tapers is None or tapers.marginal is None:
+        chol_S = tria(jnp.concatenate([y_dev.T / jnp.sqrt(N - 1), chol_R], axis=1))
+    else:
+        # Not a straightforward way to compute this via tria
+        # because tapers.marginal cannot be applied to y_dev.T / sqrt(N-1)
+        # directly. So we must compute the Cholesky factor directly.
+        C_yy = tapers.marginal * (y_dev.T @ y_dev / (N - 1))
+        chol_S = jnp.linalg.cholesky(C_yy + chol_R @ chol_R.T)
```
Comment (AdrienCorenflos) [inline cuthbertlib/ensemble_kalman/filtering.py]: Aouch, a cholesky! Can we discuss this before it goes in? Can you write the actual equations so we see if we can avoid the cholesky

Comment (DanWaxman) [reply cuthbertlib/ensemble_kalman/filtering.py]: I know I know I was sad I couldn't figure it out :(. The actual equations are:

$$ C_{yy} = \rho_{yy} \odot \sum_n (\hat{y}^{(n)}_{t | t-1} - \bar{\hat{y}}_{t | t-1}) (\hat{y}^{(n)}_{t | t-1} - \bar{\hat{y}}_{t | t-1})^\top / N-1 ,$$

where $\bar{\hat{y}}_{t | t-1}$ is the mean over the ensemble. The ensemble Kalman gain used later is then 

$$ K = C_{xy} \underbrace{(C_{yy} + R)}_{S}{}^{-1}, $$

which is used in the update

$$ x^{(n)}_{t|t} = x^{(n)}_{t|t-1} + K (y_t - \hat{y}_{t|t-1}). $$

The issue is then that we must get a factor of $S.$ I didn't see an obvious way to get a factor of $S = C_{yy} + R$ unless the taper $\rho_{yy}$ comes factored, i.e., $\rho_{yy} = L_\rho L_\rho^\top$. 

It's worth noting that I don't anticipate hitting this branch very often. It only happens if a _marginal_ covariance taper is specified. To the best of my knowledge, it is more common to just supply a _cross_ covariance taper, which avoids a Cholesky call.

File: cuthbertlib/ensemble_kalman/filtering.py (hunk 1 above)
Comment (DanWaxman) [reply cuthbertlib/ensemble_kalman/filtering.py]: Sure, I like this idea! One thing that I'm a bit unclear on is how to do this for cross vs. marginal. In particular, I think the marginal taper requires a direct Cholesky decomposition (unless Adrien has any ideas), but cross-covariance-only tapering can avoid this. 

One option is to just write `modify_cross_covariance` and `modify_marginal_covariance` separately?

Another option would be to use a `ModifiedCovariance` return type that stores `modified_cross_cov : Array` and `modified_marginal_cov : Array | None`. I think I like the former better, but can also see merits in the latter.

Comment (SamDuffield) [reply cuthbertlib/ensemble_kalman/filtering.py]: Former is better: more functional and more transparent!

Comment (DanWaxman) [reply cuthbertlib/ensemble_kalman/filtering.py]: Updated! Relevant L96 example code now looks like (streamlined a little bit from the example notebook, which is slightly more general):

```python
def periodic_distance(left, right):
    """Return pairwise shortest distances on the Lorenz–96 ring."""
    direct = jnp.abs(left[:, None] - right[None, :])
    return jnp.minimum(direct, state_dim - direct)

cross_distances = periodic_distance(state_locations, observation_locations)
marginal_distances = periodic_distance(observation_locations, observation_locations)

def modify_cross_covariance(C_xy, model_inputs):
    return gaspari_cohn(cross_distances, support_radius) * C_xy

def modify_marginal_covariance(C_yy, model_inputs):
    return gaspari_cohn(marginal_distances, support_radius) * C_yy

filter_obj = ensemble_kalman_filter.build_filter(
        init_sample=init_sample,
        get_dynamics=get_dynamics,
        get_observations=get_observations,
        n_particles=n_members,
        inflation=inflation,
        perturbed_obs=True,
        modify_cross_covariance=modify_cross_covariance,
        modify_marginal_covariance=modify_marginal_covariance,
    )
```

Passing in `modify_marginal_covariance=None` maintains the square root update. Modifiers are applied before pivoting from NaNs, so they should retain the correct indexing.

**Comment by Sahel13** [comment]
Any reason `docs/api_cuthbertlib/ensemble_kalman.md` was deleted? There a couple of links in `zensical.toml` and `docs/api_cuthbertlib/index.md` still pointing to this.

**Review by Sahel13 — APPROVED** [review]
The docs can be cleaned up later as well, there are some other minor issues that also need to be fixed unrelated to the ensemble Kalman

File: cuthbertlib/ensemble_kalman/filtering.py (hunk 2 above)
Comment (AdrienCorenflos) [reply cuthbertlib/ensemble_kalman/filtering.py]:
```
@jax.jit
def compute_enkf_cholesky_factor_qr(Y, L_rho, L_R):
    """
    Computes the lower-triangular Cholesky factor of S = rho_yy * (Y @ Y.T) + R
    using an augmented matrix QR formulation to avoid explicit covariance construction.
    
    Parameters
    ----------
    Y : jax.numpy.ndarray
        Normalized ensemble perturbations array of shape (dy, N).
    L_rho : jax.numpy.ndarray
        Lower-triangular Cholesky factor of the tapering matrix, shape (dy, dy).
    L_R : jax.numpy.ndarray
        Lower-triangular Cholesky factor of the observation noise matrix, shape (dy, dy).
        
    Returns
    -------
    L_S : jax.numpy.ndarray
        Lower-triangular Cholesky factor of S, shape (dy, dy).
    """
    dy, N = Y.shape
    
    # Construct $\tilde{Y} = [\text{diag}(l_1)Y \mid \dots \mid \text{diag}(l_{d_y})Y]$ 
    # L_rho[:, :, None] -> shape (dy, dy, 1)
    # Y[:, None, :]     -> shape (dy, 1, N)
    # Z                 -> shape (dy, dy, N)
    # Y_tilde           -> shape (dy, dy * N)
    Z = L_rho[:, :, None] * Y[:, None, :]
    Y_tilde = Z.reshape((dy, dy * N))

    # Construct the augmented matrix A = [L_R | Y_tilde]
    A = jnp.hstack([L_R, Y_tilde])

    # Compute QR decomposition of A^T. A^T is of shape (dy + N*dy, dy).
    R_qr = jnp.linalg.qr(A.T, mode='r')
    L_S_qr = R_qr.T

    # Enforce positive diagonal elements for uniqueness
    sign_diag = jnp.sign(jnp.diagonal(L_S_qr))
    # Handle zeros to avoid zeroing out entire columns
    zero_flag = (sign_diag < 0.5) & (sign_diag > -0.5) 
    sign_diag = jnp.where(zero_flag, jnp.ones_like(sign_diag), sign_diag)
    L_S_qr = L_S_qr * sign_diag[None, :]
    
    # Clean up tiny floating point noise above the diagonal
    L_S_qr = jnp.tril(L_S_qr)
    
    return L_S_qr
 ```

Comment (AdrienCorenflos) [reply cuthbertlib/ensemble_kalman/filtering.py]: @DanWaxman that should do the trick, I tested on some cases, trusting you to check properly :D

**Comment by DanWaxman** [comment]
Thanks! I updated the docs (I got a bit confused, was using the structure for `docs/cuthbert/ensemble_kalman` from #271 that I don't think made it over fully to cuthbertlib). Should be fixed now. I'll take a look at Adrien's comment now!

File: cuthbertlib/ensemble_kalman/filtering.py (hunk 2 above)
Comment (AdrienCorenflos) [reply cuthbertlib/ensemble_kalman/filtering.py]: Here 

$$ Y = (\hat{y}^{(n)}_{t | t-1} - \bar{\hat{y}}_{t | t-1}) /\sqrt{N-1} $$

Btw, if this works can you add me as a co-author in the commit so I appear on the blame?

Co-authored-by: AdrienCorenflos <adrien.corenflos@gmail.com>

same as in https://docs.github.com/en/pull-requests/how-tos/commit-changes/creating-a-commit-with-multiple-authors

Comment (DanWaxman) [reply cuthbertlib/ensemble_kalman/filtering.py]: Thanks Adrien! I agree that this works. I guess two things from there:

1. I was trying to avoid assuming that $\rho_{yy}$ had a computed Cholesky factor. We can compute it, but this additionally assumes that the taper is PSD/well-conditioned. This is probably a generally reasonable assumption, though I can imagine places where this runs into numerical trouble.
2. Sam had suggested moving to a more general `marginal_covariance_modifier(C_yy)` API. I quite like this, as it is more general (and there are some things you may want to implement that are only possible this way; for example, [shrinkage-based estimators](https://arxiv.org/abs/1502.00301)). Using the above form would necessitate moving back to the taper-only form (or a more complicated API for the user). I don't have very strong feelings here, but it does seem relevant to talk through.

**Review by Sahel13 — CHANGES_REQUESTED** [review]
I think it's best to wait to get @SamDuffield's opinion on this

File: cuthbertlib/ensemble_kalman/filtering.py (hunk 2 above)
Comment (SamDuffield) [reply cuthbertlib/ensemble_kalman/filtering.py]: I'm happy with requiring $\rho_{yy}$ in Cholesky form, if it becomes a problem later we can address it.

Regarding 2, yeah I think the functional form would be good. Maybe we can expand the input function to be `construct_chol_cross_covariance(Y, L_R)` for ensemble perturbations $Y$ and Cholesky factor of R in Adrien's notation. Then the default can be Adrien's function with `L_rho=jnp.eye`. What do you think?

Comment (DanWaxman) [reply cuthbertlib/ensemble_kalman/filtering.py]: > Then the default can be Adrien's function with L_rho=jnp.eye. What do you think?

Hmm, I think the drawback here is requiring a much larger `tria` solve if the user isn't actually doing tapering. I would probably prefer keeping a `None` path that reflects the current code structure, just using `tria` instead of `chol` for the cross-covariance-modified branch?

Comment (SamDuffield) [reply cuthbertlib/ensemble_kalman/filtering.py]: oh yes good point, for sure!

Comment (DanWaxman) [reply cuthbertlib/ensemble_kalman/filtering.py]: Updated this! 

> This commit adds use of generalized Cholesky factors under observation-space localization. This works by creating a general hook for creating an innovation covaraince factor, of which, tapering is a special case. The tapering special case is given as an augmented `tria` call in `construct_tapered_chol_innovation_covariance`, assuming that the taper matrix has a Cholesky factor.
> 
> The form of the tapered solve is due to Adrien Corenflos (see https://github.com/state-space-models/cuthbert/pull/273#discussion_r3821592615).
> 
> The branching logic in the EnKF remains unchanged, i.e., allows a branch that skips localization, as (i) observation-space tapering here implies a larger-dimensional QR solve and (ii) missingness requires an unnecessary Cholesky pivot if a default is provided.

File: cuthbertlib/enkf/smoothing.py
Comment (SamDuffield) [inline cuthbertlib/enkf/smoothing.py:1]: Something going on with git here? Don't think this file should be added

File: cuthbertlib/ensemble_kalman/filtering.py [hunk 3]
```diff
@@ -17,6 +17,25 @@
 
 ObservationFn = Callable[[Array], Array]
 DynamicsFn = Callable[[Array, KeyArray], Array]
+CrossCovarianceModifier = Callable[[Array], Array]
+ConstructLocalizedCholInnovationCovariance = Callable[[Array, Array], Array]
+
+
+def no_covariance_modifier(covariance: Array, *args: Any, **kwargs: Any) -> Array:
```
Comment (SamDuffield) [inline cuthbertlib/ensemble_kalman/filtering.py]: Do we need `args` and `kwargs` here?

File: cuthbertlib/ensemble_kalman/filtering.py [hunk 4]
```diff
@@ -56,6 +75,9 @@ def update(
     chol_R: Array,
     y: Array,
     perturbed_obs: bool = True,
+    cross_covariance_modifier: CrossCovarianceModifier = no_covariance_modifier,
+    construct_localized_chol_innovation_covariance: ConstructLocalizedCholInnovationCovariance
```
Comment (SamDuffield) [inline cuthbertlib/ensemble_kalman/filtering.py]: I would maybe rename this `construct_chol_innovation_covariance` since there is no requirement that is has to be localised, this is only an obvious use case, but there may be others

File: cuthbertlib/ensemble_kalman/filtering.py [hunk 5]
```diff
@@ -71,6 +93,13 @@ def update(
         y: Observation vector, shape (y_dim,). NaNs indicate missing dimensions.
         perturbed_obs: If True, use perturbed observations (stochastic EnKF).
             If False, use deterministic update.
+        cross_covariance_modifier: Function that modifies the empirical
+            state-observation cross-covariance. Defaults to the identity.
+        construct_localized_chol_innovation_covariance: Optional function that
```
Comment (SamDuffield) [inline cuthbertlib/ensemble_kalman/filtering.py]: Would state clearly in the docs here what the input and output are for `construct_localized_chol_innovation_covariance` or if you like consider a `Protocol` rather than `Callable` to make it clear

File: cuthbertlib/ensemble_kalman/filtering.py (hunk 3 above)
Comment (DanWaxman) [reply cuthbertlib/ensemble_kalman/filtering.py]: It's for typing, otherwise it doesn't conform to the `ModifyCrossCovariance` `Protocol`.

File: cuthbertlib/enkf/smoothing.py
Comment (DanWaxman) [reply cuthbertlib/enkf/smoothing.py:1]: Erm yeah, weird, sorry. Not sure if I accidentally popped an old stash or something. Fixed.

File: cuthbertlib/ensemble_kalman/filtering.py (hunk 4 above)
Comment (DanWaxman) [reply cuthbertlib/ensemble_kalman/filtering.py]: Sure, done.

File: cuthbertlib/ensemble_kalman/filtering.py (hunk 3 above)
Comment (SamDuffield) [reply cuthbertlib/ensemble_kalman/filtering.py]: Really? I'm only seeing `CrossCovarianceModifier = Callable[[Array], Array]` which would actually require no `args` or `kwargs`. Am I missing something?

Comment (SamDuffield) [reply cuthbertlib/ensemble_kalman/filtering.py]: Oh I see it's with the `cuthbert` interface, hmm we should definitely be able to do this without needing `args`, `kwargs`

Comment (SamDuffield) [reply cuthbertlib/ensemble_kalman/filtering.py]: For example I'd be happy to have two default functions for `no_covariance_modifier` in both `cuthbert` and `cuthbertlib`. They can even be called `_no_covariance_modifier` or similar so they don't need docstrings. I just don't like the `args`, `kwargs` 😄

File: cuthbertlib/ensemble_kalman/filtering.py (hunk 5 above)
Comment (DanWaxman) [reply cuthbertlib/ensemble_kalman/filtering.py]: I documented this in the docstring; I thought about more protocols, but they differ between cuthbert and cuthbertlib (the former takes in model inputs), and this seemed more easily confused.

**Review by SamDuffield — APPROVED** [review]
Happy with this, thanks a lot @DanWaxman! As always, if in future we find it's not suitably general for some use cases we can reevaluate

Think we also need @Sahel13 to approve as he is currently on Changes requested

**Review by Sahel13 — APPROVED** [review]

File: cuthbertlib/ensemble_kalman/filtering.py (hunk 2 above)
Comment (AdrienCorenflos) [reply cuthbertlib/ensemble_kalman/filtering.py]: I hadn't seen the rest of the convo. The code doesn't assume PSD, just that you have a triangular representation (cholesky factor is just what we call it). I think that should be ok to require?
